consent/strategy_default.go

```diff
@@ -1178,3 +1214,136 @@ func (s *DefaultStrategy) loginSessionFromCookie(r *http.Request) *flow.LoginSes
 
 	return ls
 }
+
+func (s *DefaultStrategy) requestDevice(ctx context.Context, w http.ResponseWriter, r *http.Request, req fosite.Requester) error {
+	return s.forwardDeviceRequest(ctx, w, r, req)
+}
+
+func (s *DefaultStrategy) forwardDeviceRequest(ctx context.Context, w http.ResponseWriter, r *http.Request, req fosite.Requester) error {
```

Comment:
Sorry, this is confusing me. Why do we generate a device challenge and a device verifier? All we should require is the `user_code` IMHO.

Reply:
https://www.rfc-editor.org/rfc/rfc8628

Ill need to refresh myself on this (its been a while) but the RFC states that both a user_code and a device_code be generated. 

One embedded into URL that is show to the user (either a link / QR code ect) and the other is used by the device to poll the authentication server to ascertain if the session / login has passed & accepted

Reply:
Correct, but the `device_code` is used against the token endpoint. Here, we are in the handler that the user calls, and we issue a `device_verifier` and `device_challenge`. This is probably inspired from the `{login,consent}_{verifier,challenge}`, but not really needed here, because supplying a correct `user_code` already suffices, if we store all related information in the DB and as indexable by this code.

Reply:
Ah, I think I resolved my confusion. You generate the device challenge and verifier in order to force the user to confirm the `user_code` that is associated with the particular flow. Because if the user does not need to confirm the code, an attacker could generate a link to authorize a device under the attacker's control.

Other implementations mention a confirm step after consent, but I think it is even better how it is set up here. I looked at https://github.com/ory/hydra-login-consent-node/compare/master...supercairos:hydra-login-consent-node:master and it makes sense.

Do you think it makes sense to also have a `/admin/oauth2/auth/requests/device/reject` endpoint to allow the user to abort the process? This would come into play if the codes displayed on the device do not match with the code displayed in the UI, or if the user otherwise does not recognize the code (phishing attack). Rejecting would then also respond `access_denied` to the device polling the token endpoint. WDYT?